File: clients/dotnet/Booblik/Producer.cs
using System.Diagnostics;
using System.Threading.Channels;

namespace Booblik;

/// <summary>Tunes the accumulator.</summary>
public sealed record ProducerConfig
{
    /// <summary>Records per request. Reached first, the batch goes at once.</summary>
    public int MaxBatchSize { get; init; } = 100;

    /// <summary>How long an incomplete batch waits for company.</summary>
    /// <remarks>
    /// Zero is not the fast setting. It sends every record on its own, which the broker's own
    /// measurements put at 80 592 records/s against 4 335 482 for batches of a hundred — the
    /// accumulator is the single largest performance factor in this project, an order of magnitude
    /// more than the choice of write path. Non-zero trades a bounded amount of latency for it.
    /// </remarks>
    public TimeSpan Linger { get; init; } = TimeSpan.FromMilliseconds(5);

    public AckPolicy Ack { get; init; } = AckPolicy.Written;

    /// <summary>
    /// Bounds one delivery, so a broker that stops answering fails the records waiting on it instead
    /// of hanging the accumulator for ever.
    /// </summary>
    public TimeSpan RequestTimeout { get; init; } = TimeSpan.FromSeconds(30);
}

/// <summary>Accumulates records and sends them in batches.</summary>
/// <remarks>
/// <para>
/// <b>It owns its Connection</b>: one loop holds the pending records and is the only writer to that
/// socket. Do not use the same Connection directly while a Producer has it — responses are matched
/// in order, and a second writer takes somebody else's answer.
/// </para>
/// <para>
/// Records for different partitions accumulate separately and go out as separate requests, because
/// a request addresses one partition — a partition being what has one writer.
/// </para>
/// </remarks>
public sealed class Producer : IAsyncDisposable
{
    /// <summary>
    /// What a send completes with when the batch went out under <see cref="AckPolicy.None"/>. The
    /// record was sent; no offset exists, because none is assigned until the writer reaches the batch.
    /// </summary>
    public const long OffsetUnknown = -1;

    private readonly Connection _connection;
    private readonly ProducerConfig _config;
    private readonly Channel<Command> _mailbox = Channel.CreateUnbounded<Command>();
    private readonly Task _loop;
    private int _closed;

    public Producer(Connection connection, ProducerConfig? config = null)
    {
        _connection = connection;
        _config = config ?? new ProducerConfig();
        _loop = Task.Run(RunAsync);
    }

    /// <summary>Queues a record and returns where its offset will arrive.</summary>
    /// <remarks>
    /// The record is not on the wire when this returns — that is the point. Await the task to know
    /// it landed, or call <see cref="FlushAsync"/> to push everything queued.
    /// </remarks>
    public Task<long> SendAsync(string topic, int partition, byte[] record)
    {
        ObjectDisposedException.ThrowIf(Volatile.Read(ref _closed) != 0, this);

        var completion = new TaskCompletionSource<long>(TaskCreationOptions.RunContinuationsAsynchronously);
        if (!_mailbox.Writer.TryWrite(new Command(topic, partition, record, completion, null)))
        {
            throw new InvalidOperationException("booblik: producer is closed");
        }

        return completion.Task;
    }

    /// <summary>Sends everything queued and waits for the broker to answer all of it.</summary>
    public async Task FlushAsync()
    {
        var done = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
        if (!_mailbox.Writer.TryWrite(new Command(string.Empty, 0, [], null, done)))
        {
            throw new InvalidOperationException("booblik: producer is closed");
        }

        await done.Task.ConfigureAwait(false);
    }

    /// <summary>Flushes what is queued and stops the accumulator. Does not close the Connection.</summary>
    public async ValueTask DisposeAsync()
    {
        if (Interlocked.Exchange(ref _closed, 1) != 0)
        {
            return;
        }

        _mailbox.Writer.Complete();
        await _loop.ConfigureAwait(false);
    }

    private async Task RunAsync()
    {
        var pending = new Dictionary<(string Topic, int Partition), Batch>();
        var window = new Stopwatch();
        var reader = _mailbox.Reader;

        while (true)
        {
            using var timer = pending.Count > 0 ? new CancellationTokenSource(Remaining(window)) : null;

            bool available;
            try
            {
                // `WaitToReadAsync` and then `TryRead`, never `ReadAsync(token)`. Waiting only
                // signals that something is there and takes nothing, and `TryRead` is synchronous —
                // so a cancelled window cannot take a record off the channel and then drop it. That
                // is exactly what the equivalent on the JVM did twice, after which whoever awaited
                // that record's offset waited for ever while the accumulator served everybody else.
                available = await reader.WaitToReadAsync(timer?.Token ?? CancellationToken.None)
                    .ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                await DeliverAndResetAsync().ConfigureAwait(false);
                continue;
            }

            if (!available)
            {
                await DeliverAndResetAsync().ConfigureAwait(false);
                return;
            }

            while (reader.TryRead(out var command))
            {
                if (command.Flush is { } flush)
                {
                    await DeliverAndResetAsync().ConfigureAwait(false);
                    flush.TrySetResult();
                    continue;
                }

                var key = (command.Topic, command.Partition);
                if (!pending.TryGetValue(key, out var batch))
                {
                    batch = new Batch();
                    pending[key] = batch;

                    // The window is measured from the **first** record of the batch and never
                    // restarted. Timing from the last would let a steady trickle postpone the send
                    // indefinitely, turning a latency bound into a latency hope.
                    if (!window.IsRunning)
                    {
                        window.Restart();
                    }
                }

                batch.Records.Add(command.Record);
                batch.Waiting.Add(command.Completion!);

                if (batch.Records.Count >= _config.MaxBatchSize)
                {
                    await DeliverAndResetAsync().ConfigureAwait(false);
                }
            }
        }

        // Delivering always stops the window. Without this the stopwatch keeps running past the
        // send, and the next batch is born with its window already spent — a linger setting that
        // works once and then behaves as zero for the rest of the process.
        async Task DeliverAndResetAsync()
        {
            await DeliverAsync(pending).ConfigureAwait(false);
            window.Reset();
        }

        TimeSpan Remaining(Stopwatch running)
        {
            var left = _config.Linger - running.Elapsed;
            return left > TimeSpan.Zero ? left : TimeSpan.Zero;
        }
    }

    private async Task DeliverAsync(Dictionary<(string Topic, int Partition), Batch> pending)
    {
        if (pending.Count == 0)
        {
            return;
        }

        foreach (var (key, batch) in pending.ToArray())
        {
            pending.Remove(key);

            using var timeout = new CancellationTokenSource(_config.RequestTimeout);
            try
            {
                var result = await _connection
                    .ProduceAsync(key.Topic, key.Partition, batch.Records, _config.Ack, timeout.Token)
                    .ConfigureAwait(false);

                for (var index = 0; index < batch.Waiting.Count; index++)
                {
                    // One request is written by one call, so the records are contiguous.
                    batch.Waiting[index].TrySetResult(
                        result is { } offsets ? offsets.BaseOffset + index : OffsetUnknown);
                }
            }
            catch (Exception failure)
            {
                // A batch that fails fails for all of its records. Completing some and abandoning
                // the rest would leave callers awaiting a task nothing will ever finish.
                foreach (var waiting in batch.Waiting)
                {
                    waiting.TrySetException(failure);
                }
            }
        }
    }

    private sealed record Command(
        string Topic,
        int Partition,
        byte[] Record,
        TaskCompletionSource<long>? Completion,
        TaskCompletionSource? Flush);

    private sealed class Batch
    {
        public List<byte[]> Records { get; } = [];

        public List<TaskCompletionSource<long>> Waiting { get; } = [];
    }
}

File: clients/python/booblik/producer.py
"""The accumulator.

One record per request throws away the largest single performance factor in this project: the
broker's own measurements put batches of a hundred at 4 335 482 records/s against 80 592 one at a
time. That is why batching is not an optimisation to enable later — it is what a producer is.
"""

from __future__ import annotations

import queue
import threading
import time
from concurrent.futures import Future
from dataclasses import dataclass, field

from .connection import AckPolicy, Connection

#: What a Future carries when the batch went out under :attr:`AckPolicy.NONE`. The record was sent;
#: no offset exists, because none is assigned until the writer reaches the batch.
OFFSET_UNKNOWN = -1


@dataclass
class ProducerConfig:
    #: Records per request. Reached first, the batch goes at once.
    max_batch_size: int = 100
    #: How long an incomplete batch waits for company, in seconds.
    #:
    #: Zero is not the fast setting — it sends every record on its own. Non-zero trades a bounded
    #: amount of latency for the factor above.
    linger: float = 0.005
    ack: AckPolicy = AckPolicy.WRITTEN
    #: Bounds one delivery, so a broker that stops answering fails the records waiting on it
    #: instead of hanging the accumulator for ever.
    request_timeout: float = 30.0


@dataclass
class _Append:
    topic: str
    partition: int
    record: bytes
    future: Future


@dataclass
class _Flush:
    done: threading.Event


@dataclass
class _Batch:
    records: list[bytes] = field(default_factory=list)
    futures: list[Future] = field(default_factory=list)


_CLOSE = object()


class Producer:
    """Accumulates records and sends them in batches.

    **It owns its Connection**: one thread holds the pending records and is the only writer to that
    socket. Do not use the same Connection directly while a Producer has it — responses are matched
    in order, and a second writer takes somebody else's answer.

    Records for different partitions accumulate separately and go out as separate requests, because
    a request addresses one partition — a partition being what has one writer.
    """

    def __init__(self, connection: Connection, config: ProducerConfig | None = None):
        self._connection = connection
        self._config = config or ProducerConfig()
        self._mailbox: queue.Queue = queue.Queue()
        self._closed = threading.Event()
        self._thread = threading.Thread(target=self._run, name="booblik-producer", daemon=True)
        self._thread.start()

    def send(self, topic: str, partition: int, record: bytes) -> Future:
        """Queues a record and returns where its offset will arrive.

        The record is not on the wire when this returns — that is the point. Call ``result()`` on
        the future to know it landed, or :meth:`flush` to push everything queued.
        """
        if self._closed.is_set():
            raise RuntimeError("producer is closed")
        future: Future = Future()
        self._mailbox.put(_Append(topic, partition, record, future))
        return future

    def flush(self, timeout: float | None = None) -> None:
        """Sends everything queued and waits for the broker to answer all of it."""
        done = threading.Event()
        self._mailbox.put(_Flush(done))
        if not done.wait(timeout):
            raise TimeoutError("flush timed out")

    def close(self) -> None:
        """Flushes what is queued and stops the accumulator. Does not close the Connection."""
        if self._closed.is_set():
            return
        self._closed.set()
        self._mailbox.put(_CLOSE)
        self._thread.join()
        self._fail_leftovers()

    def _fail_leftovers(self) -> None:
        """Fails anything that reached the mailbox after the loop had already returned.

        A `send` that passed the closed check just before `close` ran can put its record after the
        loop is gone, and it would then wait on a future nothing will ever complete. Failing those
        is the difference between a caller that learns something and a caller that stops.

        The Go client had the same race with a worse ending — it closed its mailbox, so the late
        send panicked outright, and that is how this was noticed at all. Here the window is narrow
        enough that threads do not reliably hit it, which is why the test calls this directly
        instead of pretending to reproduce a race it cannot.
        """
        while True:
            try:
                leftover = self._mailbox.get_nowait()
            except queue.Empty:
                return
            if isinstance(leftover, _Append):
                leftover.future.set_exception(RuntimeError("producer is closed"))
            elif isinstance(leftover, _Flush):
                leftover.done.set()

    def __enter__(self) -> "Producer":
        return self

    def __exit__(self, *_) -> None:
        self.close()

    def _run(self) -> None:
        pending: dict[tuple[str, int], _Batch] = {}
        # When the current window ends, or None when nothing is waiting.
        deadline: float | None = None

        while True:
            timeout = None if deadline is None else max(0.0, deadline - time.monotonic())
            try:
                # `get` with a timeout either returns an item or raises Empty. It cannot take an
                # item off the queue and then drop it, which is what the equivalent on the JVM did
                # twice — a timeout there cancels the pending receive, and a cancelled receive could
                # swallow a record, after which whoever awaited its offset waited for ever.
                command = self._mailbox.get(timeout=timeout)
            except queue.Empty:
                self._deliver(pending)
                deadline = None
                continue

            if command is _CLOSE:
                self._deliver(pending)
                return

            if isinstance(command, _Flush):
                self._deliver(pending)
                deadline = None
                command.done.set()
                continue

            batch = pending.setdefault((command.topic, command.partition), _Batch())
            batch.records.append(command.record)
            batch.futures.append(command.future)

            # The window is measured from the **first** record of the batch and never restarted.
            # Timing from the last would let a steady trickle postpone the send indefinitely,
            # turning a latency bound into a latency hope.
            if deadline is None:
                deadline = time.monotonic() + self._config.linger
            if len(batch.records) >= self._config.max_batch_size:
                self._deliver(pending)
                deadline = None

    def _deliver(self, pending: dict[tuple[str, int], _Batch]) -> None:
        for (topic, partition), batch in list(pending.items()):
            del pending[(topic, partition)]

            try:
                result = self._connection.produce(topic, partition, batch.records, self._config.ack)
            except Exception as failure:  # noqa: BLE001 — every waiting caller has to learn of it
                for future in batch.futures:
                    future.set_exception(failure)
                continue

            for index, future in enumerate(batch.futures):
                if result is None:
                    future.set_result(OFFSET_UNKNOWN)
                else:
                    # One request is written by one call, so the records are contiguous.
                    future.set_result(result.base_offset + index)
